### DISCUSSION TITLE
How to only get versions of a master of a certain format

### DISCUSSION THREAD

**Discussion by Lightdm** [joalla/discogs_client#155] [Support] [answered]
Hi everyone. I would like to only get versions of a master matching certain parameters, like the format. As far as I can see the only way to do this using this repo is fetching all versions of the master
```
master = d.master(ID)
versions = master.versions
```
and then going through all these versions by hand, for each one queuing `version.formats[0]["name"] == "Vinyl"` (when checking for the format to by vinyl). This of course takes a long time and a lot of API requests.
The Discogs API seems to already support only fetching versions conforming to some filter ([Docs](https://www.discogs.com/developers#page:database,header:database-master-release-versions)).
I did not find a way to include this filter using this python client, is it possible, and I just didn't find it, or has this not been added?

**Comment by JOJ0** [comment]
Unfortunately I think that's not supported. We only can use this api https://www.discogs.com/developers#page:database,header:database-search

it can filter for format but it only searches releases not masters :(

Hope I got this right on a quick look...

I'll convert this to a feature request but can't promise anybody will pick it up soon.

**Reply by JOJ0** [reply]
but wait, there is a type filter (and format too) in the search api, maybe it does what you want:


```
type
string (optional) Example: release
String. One of release, master, artist, label
```

here's an example how to use search api with parameters: https://python3-discogs-client.readthedocs.io/en/latest/quickstart.html#searching

**Reply by Lightdm** [reply]
Hmm, I am talking specifically about this part of the Discogs Docs: https://www.discogs.com/developers#page:database,header:database-master-release-versions
There they explain how to get the master via `/masters/{master_id}` and the versions of a master via `/masters/{master_id}/versions`. This GET request (`/masters/{master_id}/versions`), seems to be capable of passing on a desired format 
![Image showing the format as an accepted parameter](https://github.com/user-attachments/assets/79c12438-59b9-4e6e-b5c0-a72355ddfc77)
I did not fully understand the source code yet, I only found the call to fetch the master itself [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/models.py#L608), but I did not find where the Master.versions is populated.
Thanks for your quick answer!

**Reply by AnssiAhola** [reply, marked as answer]
It's pretty hard to figure it out from the source code, but `Master.versions` is actually an `PaginatedList` under the hood which inherits [`BasePaginatedResponse`](https://github.com/joalla/discogs_client/blob/master/discogs_client/models.py#L268) so it has the `filter()` method and should call that specific api when invoked.

Example
```python
master = client.master(1051980)
result = master.versions.filter(format="CD")
for release in result:
    print(release)
```
The `filter()` method supports all the parameters described in the API docs and can be passed in as kwargs
```python
result = master.versions.filter(
    master_id=123,
    page=1,
    per_page=10,
    format="Vinyl",
    label="Scorpio Music",
    released=1992,
    country="Belgium",
    sort="released",
    sort_order="asc"
)
```

Hope that helps, Cheers!

**Reply by Lightdm** [reply]
That is very much exactly what I was looking for, thank you very very much! I stumbled upon filter when I searched through the source code, but couldn't get it to work, thank you for showing me!

**Reply by JOJ0** [reply]
OMG thanks @AnssiAhola the saviour as usual!!! Sorry for being misleading here @Lightdm 😅
